### PR TITLE
Parallelize end-to-end tests

### LINKED ISSUE #2646: Reduce execution time of end-to-end tests
They currently take ~ 20 mins , which is a long time.

We could parallelize them by:
* Running multiple tests in parallel in the same cluster. This causes conflicts between the tests. For instance it would require ensure that the resources created in each test are unique, to avoid clashes. 
* Using multiple kubernetes clusters and distributing tests across the clusters, to be run sequentially in each cluster. This may make it harder to run tests in pre-existing clusters (e.g. locally) but I guess we can always add an option to run tests sequentially or use a single cluster.

### PR DESCRIPTION
**Pull request by 2opremio**
The parallelization happens by running tests in multiple Kubernetes clusters (created with Kind).

I've had to fork bats in order to wire the clusters with the tests. I am waiting on https://github.com/bats-core/bats-core/pull/255 to be merged before using upstream-bats again.

Let's see what results we get in CircleCI ...

Fixes #2646

### CODE REVIEW

**Comment by 2opremio** [comment]
With two jobs the end to end tests take 11:00 minutes instead of [16:40](https://circleci.com/gh/fluxcd/flux/8781) which is a 44% time reduction!

Let's see if things keep improving with 3 clusters or the overhead is too high.

**Comment by 2opremio** [comment]
3 clusters doesn't really make it faster (it takes 11 minutes all the same), so let's keep it at 2.

**Comment by 2opremio** [comment]
It turns out that the free tier of CircleCI _only_ provides two CPU cores: https://circleci.com/docs/2.0/configuration-reference/#resource_class

That explains it, let's see if I can make the switch to large.

**Comment by 2opremio** [comment]
I created a ticket at the CNCF support desk to upgrade the fluxcd org to CircleCI's performance plan. 

However, I just realized that there is a per-user payment, which doesn't really make sense in an open source project with external contributors, so I will clarify this with CircleCI first. I just created a support ticket with them to.

**Comment by 2opremio** [comment]
I just left a call with some CircleCI guys and it turns out that the per-user charge only applies to users making use of private repos (which we have none).

The CircleCI stats are currently broken but it seems we are using 50K credits per month. Payment goes in blocks of 25K creates so it shouldn't translate in more than $30 per month.

I am looking into getting into a trial.

**Comment by 2opremio** [comment]
It seems we weren't put on the trial plan yet. I am in conversations with the CNCF helpdesk so that the upgrade the plan.

**Comment by 2opremio** [comment]
Let's get this merged first and then change the number of clusters once we are upgraded to the performance plan.

**Comment by 2opremio** [comment]
I have now upgraded the e2e tests and CircleCI to 4 cores, which brings the `e2e` tests execution time to ~9 minutes (from 11 minutes with 2 cores). This is way better than the initial 16 minutes, but it's still far from the theoretical 50% improvement, which should had brought us down to 5:30. 

Most of the overhead comes from the kind clusters creation, which, although done in parallel, still takes a long time (~4 minutes). This overhead will be amortized when we add more tests.

**Comment by 2opremio** [comment]
The total execution time of build has gone down to ~13 minutes from ~22 minutes.

**Comment by 2opremio** [comment]
I think we can get even better by:
* running certain steps in the background
* enabling docker layer caching

File: test/e2e/run.bash [hunk 1]
```diff
@@ -12,34 +12,48 @@ CACHE_DIR="${FLUX_ROOT_DIR}/cache/$CURRENT_OS_ARCH"
...
   chmod +x "${FLUX_ROOT_DIR}/test/bin/kind"
-  kind create cluster --name "${KIND_CLUSTER}" --wait 5m
-  defer kind --name "${KIND_CLUSTER}" delete cluster > /dev/null 2>&1 || true
-  KUBECONFIG="$(kind --name="${KIND_CLUSTER}" get kubeconfig-path)"
-  export KUBECONFIG
-  USING_KIND=true
-  kubectl get pods --all-namespaces
+}
+
+# Create multiple Kind clusters and run jobs in parlallel?
```
[the remark this answers is not in the document]
Comment (squaremo) [reply test/e2e/run.bash]:
```suggestion
# Create multiple Kind clusters and run jobs in parallel?
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -90,7 +90,9 @@ jobs:
             sudo killall apt-get || true
 
             sudo apt-get update
-            sudo apt-get install -y git rng-tools memcached
+            sudo apt-get install -y git rng-tools memcached parallel
+            mkdir -p "$HOME/.parallel"
+            touch "$HOME/.parallel/will-cite"
```
[the remark this answers is not in the document]
Comment (2opremio) [reply .circleci/config.yml]: It's either this or invoking it all the time with `--will-cite`

It's a longstanding issue with GNU parallel. Some people [consider it nagware](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=884793).

File: test/e2e/run.bash (hunk 1 above)
[the remark this answers is not in the document]
Comment (2opremio) [reply test/e2e/run.bash]: it's a question since we are querying the env variable

**Pull request merged**